oeps/oep-0019-bp-developer-documentation.rst

```diff
@@ -0,0 +1,213 @@
+concern itself with any transitory “Notes” that a development team may use for
+project management, brainstorming, meeting notes, etc.  We also distinguish
+"Documentation" from any useful code "Comments" that a developer may choose
+to add to clarify or add rationale to a local code segment. "Comments" are also
+out of scope of this OEP.
+
+Decisions
+---------
+
+.. image:: oep-0019/developer_docs.png
+   :alt: A diagram that shows the different types of documentation and their
+    relationships.  This is further described below.
+
+Agile Documentation Needs
+~~~~~~~~~~~~~~~~~~~~~~~~~
+
+With Agile development and Clean Code, we can minimize over-documentation since
+code should be readable and self-documenting. Where we do still need developer
+documentation is:
+
+1. **Decision Records** to understand the rationale behind the code and the
+   evolution of its perceived tech debt.
+
+   * These will be captured in either: (1) the `open-edx-proposals repo`_ for
+     system-wide decisions or (2) in `local decisions`_ folders within local code
+     repos. The latter may exist in a top-level folder in a repo and also as a
+     subfolder of an individual Django app.
+
+2. **Diagrams** to provide a visual holistic perspective of the overall system 
+   or component with its boundaries, interconnections, and relationships.
+
+   * These will be stored in Github co-located with the relevant code or text.
+
+3. **How-To Documentation** with step-by-step instruction in enabling and using
+   a feature.
+
+   * These will be stored in Github co-located with the relevant code.
+
+4. **API Documentation** with information on interfaces for integrating with or
+   enhancing the platform. These should be automatically generated, as possible.
+   Some examples are:
+
+   * REST API interfaces will be automatically generated using Swagger_.
+   * The Open edX Learning Exchange (OLX) format should be automatically generated.
+   * Plugin interfaces (e.g., Proctoring tool) should be automatically generated
+     using Sphinx_.
+
+.. _open-edx-proposals repo: https://github.com/edx/open-edx-proposals
+.. _local decisions: https://github.com/edx/edx-platform/tree/master/openedx/core/djangoapps/oauth_dispatch/docs/decisions
+.. _Swagger: https://swagger.io/
+.. _Sphinx: http://www.sphinx-doc.org/en/master/
+
+Tools and Location
+~~~~~~~~~~~~~~~~~~
+
+The recommended tools and locations for developer documentation are described
+below, inlined with the requirements they meet.
+
+* **Discoverable**. Users must be able to find the documentation that is
+  relevant to their needs. This calls for a high-level index and search
+  mechanism for finding content. 
+
+  * **Search**. For the time-being, we will make use of Google and Github's
+    search functionality for finding content since all developer documentation
+    will be in Github.  Certain IDEs may also have good tools for searching. 
```

Comment:
The *source* for documentation will be in GitHub.  The readable and usable versions might be elsewhere, like readthedocs.

Reply:
Yes.  Does that need to be further clarified or is that already clear?